### ISSUE TITLE
[RFC] New constraints for Symfony validator

### ISSUE THREAD

**Issue by javiereguiluz** [symfony/symfony#28608]
Given that Symfony's validator is based on Java's validator, I was checking the [Bean Validation 2.0](https://beanvalidation.org/2.0/) standard to see if they introduced new features that we could implement too.

These are the new constraints that Symfony doesn't define yet (although you can implement them indirectly with other existing constraints):

Dates:

* `@Past`
* `@PastOrPresent`
* `@Future`
* `@FutureOrPresent`

Numbers:

* `@Positive`
* `@PositiveOrZero`
* `@Negative`
* `@NegativeOrZero`

In addition, [Hibernate Validator](http://hibernate.org/validator/documentation/), the only official implementation of Bean Validator 2.0, defines other constraints:

* [@CreditCardNumber](https://docs.jboss.org/hibernate/stable/validator/api/org/hibernate/validator/constraints/CreditCardNumber.html) (it does just the Luhn algorithm check, via the [@LuhnCheckValidator](https://docs.jboss.org/hibernate/stable/validator/api/org/hibernate/validator/internal/constraintvalidators/hv/LuhnCheckValidator.html) but I guess they defined this alias to make it easier to find and understand).
* [@EAN](https://docs.jboss.org/hibernate/stable/validator/api/org/hibernate/validator/internal/constraintvalidators/hv/EANValidator.html) for barcode numbers.
* [@SafeHtml](https://docs.jboss.org/hibernate/stable/validator/api/org/hibernate/validator/internal/constraintvalidators/hv/SafeHtmlValidator.html) validates that the string does not contain malicious code. It uses an external sanitizer library.

Are we interested in implementing any of these constraints?

**Comment by stof** [comment]
if `@SafeHtml` requires a third-party sanitizer library, I suggest keeping it out of Symfony. A separate composer package can provide it, and have a requirement on the library it uses (instead of dealing with optional requirements).

regarding the CreditCardNumber alias, I would point you to the discussion on https://github.com/symfony/symfony/issues/26436, which seems to tell me we don't want to make it easier to discover that a credit card number is using luhn.

Regarding EAN, this is just a Length constraint with only 2 possible allowed values for the length.

**Comment by javiereguiluz** [comment]
About the EAN validator, the last digit is a checksum digit ... so there's something to validate there. See https://gist.github.com/damonjones/7566978

**Comment by stof** [comment]
Well, but there could be indeed. But Hibernate does not, making this constraint quite useless.

**Comment by jordscream** [comment]
@javiereguiluz 

1- for the Date Validator, do we have always a interest to keep DateTimeValidator and DateValidator ? Should we keep just one ? DateValidator with format accepted by date() ?

2- Past, Future can be options of DateConstraint ? or should it be annotation using the validation of DateValidator ?

**Comment by stof** [comment]
Well, these should not be options of the Date constraints, as these constraints make sense on a property holding a DateTime object too, only on properties holding a string representing a date (which is what the Date constraint is about).

**Comment by jordscream** [comment]
so @Past.... can only work if value is instance of DateTimeInterface ?

**Comment by ro0NL** [comment]
From https://docs.jboss.org/hibernate/stable/beanvalidation/api/javax/validation/constraints/Past.html i'd say yes, but maybe we can add a format option to these to allow strings.

**Comment by ro0NL** [comment]
well.. it shows we dont need a format option

```
@Past # datetime | string
@Date @Past # string
@Type(DateTime) @Past # datetime
```

the first excepts any date string (e.g. https://3v4l.org/vk0AO), the second is narrowed to a formatted stirng

**Comment by pimolo** [comment]
About the `SafeHtml` constraint, in my opinion, if you need to make sure that a string is safe from malicious code, it should be passed into a sanitizer **anyway**.

I don't see a common use case which would require to check first if it is already safe.

**Comment by javiereguiluz** [comment]
I have an Event entity and I wanted to use https://symfony.com/blog/new-in-symfony-3-4-improved-comparison-constraints to replace my own callback validator by `@Assert\GreaterThan(propertyPath="startDate")` in the `endDate`.

However, the callback also validates that `startDate` is a future date ... and that's something you can't do with an expression validator ... so I ended up keeping the callback validator.

In summary: the `@Future` or `@FutureOrPresent` asserts would have come in handy for me today :)

**Comment by ogizanagi** [comment]
@javiereguiluz : Do you mean `@Assert\GreaterThan("today")` ? 😄 https://symfony.com/doc/current/reference/constraints/GreaterThan.html#comparing-dates

**Comment by javiereguiluz** [comment]
🤦‍♂️  Thanks Maxime 😊

**Comment by dkarlovi** [comment]
This shouldn't have been closed since it's mentioning EAN (AKA GTIN), it's quite common for developers to work with "products" and having a dedicated validator for this specific identificator seems like something you'd want. It's not just length + checksum, it also has various flavors, etc.

**Comment by norkunas** [comment]
Also that HtmlSanitizer is now proposed as new Symfony component (https://github.com/symfony/symfony/issues/44144) `@SafeHtml` constraint could be also included after :)
